### PR TITLE
Hot path for sorting planar points in btree operator class

### PR DESCRIPTION
**Pull request by Komzpa**
Removing everything that slows down btree sorting to get to maximum speed.

### CODE REVIEW

**Comment by pramsey** [comment]
Bridge too far....

**Comment by Komzpa** [comment]
This doesn't break SQL API.
https://trac.osgeo.org/postgis/ticket/3864

Further major improvements are possible, but require using sortsupport infrastructure.

**Comment by pramsey** [comment]
Can you point to sortsupport examples? I'm interested to learn more.

**Comment by Komzpa** [comment]
In docs it's called "Abbreviated keys".
Here is a structure to fill: https://github.com/postgres/postgres/blob/master/src/include/utils/sortsupport.h
Examples:
 - https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/uuid.c#L25
 - https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/varlena.c#L1382 - a big chunk of code that explains why Collate C text ordering is much faster than usual - thus why geohash sorting is faster.

**Comment by pramsey** [comment]
Not seeing any references to this in the SQL guides of PgSQL, I assume we'd have to hand enter our hooks into pg_amproc, which would be a return to old form back in the early 2000s when we had to hand enter our GIST index bindings, before the advent of CREATE OPERATOR CLASS, and friends.

**Comment by Komzpa** [comment]
@pramsey according to @funny-falcon it's doable in SQL:

function 2: https://www.postgresql.org/docs/devel/static/xindex.html#xindex-btree-support-table
example code: https://www.postgresql.org/docs/devel/static/xindex.html#xindex-opfamily

**Comment by pramsey** [comment]
Yep, just another line in the opclass, nice.

**Comment by strk** [comment]
> Yep, just another line in the opclass, nice.

Let's see if we can make it upgradeable.
IIRC I added some support for per-opclass-line upgrades
("-- Availability:" comment supported on each line)

**Comment by Komzpa** [comment]
@pramsey in IRC @robe2 said that you're the one to review this. 
I propose the following:
 - for 2.4.x, commit this;
 - for 2.5.x, do a bigger change. Add sortsupport, and discuss whether the sort needs to be on box center, or just the first point of geometry is fine too. For latter case, we can skip box construction for simple geometries that don't have it serialized, without affecting the order too much.

File: liblwgeom/g_serialized.c [hunk 1]
```diff
@@ -109,7 +110,7 @@ int32_t gserialized_get_srid(const GSERIALIZED *s)
 	if ( srid == 0 )
 		return SRID_UNKNOWN;
 	else
-		return clamp_srid(srid);
+		return srid;
```
Comment (strk) [inline liblwgeom/g_serialized.c]: This is unrelated to "clean up sorting code path", isn't it ?
Could you please rebase and make sure only related changes are seen in the "Files changed" tab on github ?

Comment (Komzpa) [reply liblwgeom/g_serialized.c]: gserialized_get_srid is one of topmost not-inlined functions in sort:
image

and it can't be inlined since it calls clamp_srid:
image

Comment (Komzpa) [reply liblwgeom/g_serialized.c]: This optimization was really walked around later, as I made srid cmp function that doesn't construct SRID. But it also doesn't clamp it, which makes `cmp_srid(a,b)` and `get_srid(a)==get_srid(b)` diverge in opinions in edge cases. I think if we can't commit to non-clamping get_srid, we can't commit to non-clamping cmp_srid, so these both should go in pair.

Since we are clamping on write, without data corruption you shouldn't get non-clamped values on disk.

Comment (strk) [reply liblwgeom/g_serialized.c]: I think the clamp-on-read was done for the restore case, when restoring a dump from pre-clamp dates (postgis_restore should take care of that, but pg_upgrade would not). Anyway this PR is about "clean up", not about "performance improvement", am I wrong ?

Comment (Komzpa) [reply liblwgeom/g_serialized.c]: Sorry for misleading wording. It is about performance, and addresses https://trac.osgeo.org/postgis/ticket/3864.

According to @robe2 pre-2.0 dumps come in HEXWKB, and for those clamping is active: https://github.com/postgis/postgis/blob/a5c7920a059c7273864760539b6845d74b34289c/liblwgeom/lwin_wkb.c#L701

**Comment by pramsey** [comment]
Committed. Further use of sortsupport fanciness should go in a fresh ticket/PR

File: liblwgeom/g_serialized.c [hunk 2]
```diff
@@ -284,8 +294,43 @@ int gserialized_cmp(const GSERIALIZED *g1, const GSERIALIZED *g2)
...
 	size_t sz2 = SIZE_GET(g2->size);
+	union floatuint x, y;
+
+	/*
+	* For two non-same planar points, we can skip a lot of machinery.
+	*/
+	if (
+		sz1 > 16 && // 16 is size of EMPTY, if it's larger - it has coordinates
+		sz2 > 16 &&
+		*(uint32_t*)(g1->data) == POINTTYPE &&
```
Comment (notorca) [inline liblwgeom/g_serialized.c]: The C Standard, 6.3.2.3, paragraph 7 [ISO/IEC 9899:2011], states
A pointer to an object or incomplete type may be converted to a pointer to a different object or incomplete type. If the resulting pointer is not correctly aligned for the referenced type, the behavior is undefined.

File: liblwgeom/g_serialized.c [hunk 3]
```diff
@@ -284,8 +294,43 @@ int gserialized_cmp(const GSERIALIZED *g1, const GSERIALIZED *g2)
...
+		*(uint32_t*)(g2->data) == POINTTYPE &&
```
Comment (notorca) [inline liblwgeom/g_serialized.c]: Same ussue here

File: liblwgeom/g_serialized.c [hunk 4]
```diff
@@ -284,8 +294,43 @@ int gserialized_cmp(const GSERIALIZED *g1, const GSERIALIZED *g2)
...
+		!FLAGS_GET_BBOX(g1->flags) &&
+		!FLAGS_GET_GEODETIC(g1->flags) &&
+		!FLAGS_GET_BBOX(g2->flags) &&
+		!FLAGS_GET_GEODETIC(g2->flags)
+	)
+	{
+		double *dptr = (double*)(g1->data);
```
Comment (notorca) [inline liblwgeom/g_serialized.c]: and here

File: liblwgeom/g_serialized.c [hunk 5]
```diff
@@ -284,8 +294,43 @@ int gserialized_cmp(const GSERIALIZED *g1, const GSERIALIZED *g2)
...
+		x.f = 2.0 * dptr[1];
+		y.f = 2.0 * dptr[2];
+		hash1 = uint32_interleave_2(x.u, y.u);
+
+		dptr = (double*)(g2->data);
```
Comment (notorca) [inline liblwgeom/g_serialized.c]: one more

**Comment by notorca** [comment]
https://www.securecoding.cert.org/confluence/display/c/EXP36-C.+Do+not+cast+pointers+into+more+strictly+aligned+pointer+types

File: liblwgeom/g_serialized.c (hunk 2 above)
Comment (pramsey) [reply liblwgeom/g_serialized.c]: The data buffer is constructed carefully so that aligned access is expected. Doubles exist on double alignment boundaries, and ints on their alignment boundary. https://github.com/postgis/postgis/blob/svn-trunk/liblwgeom/g_serialized.txt
Also, the storage type in pgSQL is defined as double aligned, so the segments we're getting back are on double-aligned boundares.
I'm not sure whether "it works" is sufficient for this objection.
One possible issue I guess is for geometries built on top of memory they allocated themselves, whether palloc() (when running within pgsql) or malloc() are returning double-aligned memory.

Comment (notorca) [reply liblwgeom/g_serialized.c]: There is no guarantee, than result of malloc have same alignment as double.

Comment (notorca) [reply liblwgeom/g_serialized.c]: And there is one more problem here, *(uint32_t*)(g1->data) and double *dptr = (double*)(g1->data); creates 2 aliases for same memory address with incompatible types.

File: liblwgeom/g_serialized.c (hunk 5 above)
Comment (notorca) [reply liblwgeom/g_serialized.c]: Should be:
```
dptr = (double*)(g2->data+sizeof(double));
x.f = 2.0 * dptr[0];
y.f = 2.0 * dptr[1];
```
